### ISSUE TITLE
rclone dedupe --fast-list does not dedupe

### ISSUE THREAD

**Issue by oqpvc** [ncw/rclone#2902]
Hi there!

I suppose that the following behaviour is unintended and should be considered a bug:

```
~ >>> rclone dedupe --dry-run --fast-list -vv Gsuite:test
2019/01/09 17:27:03 DEBUG : rclone: Version "v1.45-069-g7ee7bc87-beta" starting with parameters ["rclone" "dedupe" "--dry-run" "--fast-list" "-vv" "Gsuite:test"]
2019/01/09 17:27:03 DEBUG : Using config file from "/home/oqpvc/.rclone.conf"
2019/01/09 17:27:22 INFO  : Google drive root 'test': Looking for duplicates using interactive mode.
2019/01/09 17:27:29 INFO  : dedupetest: NOT Merging contents of duplicate directories as --dry-run
2019/01/09 17:28:07 DEBUG : 6 go routines active
2019/01/09 17:28:07 DEBUG : rclone: Version "v1.45-069-g7ee7bc87-beta" finishing with parameters ["rclone" "dedupe" "--dry-run" "--fast-list" "-vv" "Gsuite:test"]
~ >>> rclone dedupe --dry-run  -vv Gsuite:test
2019/01/09 17:28:50 DEBUG : rclone: Version "v1.45-069-g7ee7bc87-beta" starting with parameters ["rclone" "dedupe" "--dry-run" "-vv" "Gsuite:test"]
2019/01/09 17:28:50 DEBUG : Using config file from "/home/oqpvc/.rclone.conf"
2019/01/09 17:29:07 INFO  : Google drive root 'test': Looking for duplicates using interactive mode.
2019/01/09 17:29:10 INFO  : dedupetest: NOT Merging contents of duplicate directories as --dry-run
2019/01/09 17:29:30 NOTICE: dedupetest/foo.epub: Found 2 duplicates - deleting identical copies
2019/01/09 17:29:30 NOTICE: dedupetest/foo.epub: Deleting 1/2 identical duplicates (MD5 "fd35982146610378a240a276fb88c69f")
2019/01/09 17:29:30 NOTICE: dedupetest/foo.epub: Not deleting as --dry-run
2019/01/09 17:29:30 NOTICE: dedupetest/foo.epub: All duplicates removed
2019/01/09 17:29:30 DEBUG : 6 go routines active
2019/01/09 17:29:30 DEBUG : rclone: Version "v1.45-069-g7ee7bc87-beta" finishing with parameters ["rclone" "dedupe" "--dry-run" "-vv" "Gsuite:test"]
```

The directory `test` contains two directories of the same name `dedupetest`, the first containing a single file and the second containing a single file `foo.epub` twice.

I guess that somewhere in the implementation of `--fast-list` there is an implicit assumption that path names are unique identifiers, a rather silly idea if you ask whoever is in charge of Google Drive.

This makes deduplication considerably less useful: Without `--fast-list`, `rclone dedupe Gsuite:/` will reach the Google rate limit in no time.

**Comment by ncw** [comment]
> I guess that somewhere in the implementation of `--fast-list` there is an implicit assumption that path names are unique identifiers, a rather silly idea if you ask whoever is in charge of Google Drive.

Yes you are right, there is...

It would be very useful if `--fast-list` worked with dedupe.

### Ideas

What this needs is a new file iteration scheme.  dedupe uses `walk.Walk` which if used with `--fast-list` will build an in memory tree so it iterates exactly the same as `walk.Walk` without fast list.

What we should be using is something like `walk.WalkUnordered` this could then pass through the output of fast list directly.  This could use the ListR callback so it would effectively be ListR for backends supporting that.

This would be useful in `rclone size` for instance.  Perhaps `operations.ListFn` should use this iteration scheme.  This would be fine for hash/ls/lsl/lsd/delete/settier.  Maybe cat would be the only one which really cares...

Which probably means dedupe should be using ListFn too.

Note that in cmd/serve/restic/restic.go there is a workaround for exactly this issue (using ListR instead of recursive walk) and there may be other places.

**Comment by ncw** [comment]
I've made a start on this - can you have a go with

https://beta.rclone.org/branch/v1.45-107-g517486f7-fix-2902-listr-beta/ (uploaded in 15-30 mins)

In my tests it was 7x faster!  There is no need to use the `--fast-list` flag, it will use it automatically.

**Comment by oqpvc** [comment]
This looks awesome! Playing around, I can't get it to not find duplicates!

**Comment by ncw** [comment]
> This looks awesome! Playing around, I can't get it to not find duplicates!

Let me know if you find any problems!

**Comment by LingMan** [comment]
Found two folders that should get deduped on first glance but don't with this test build. I'll try to debug or at least isolate a test case in the next few days.

If I don't get around to it, take this as a heads-up that here be dragons.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.47

@LingMan if you can reproduce your problem then please let me know.

**Comment by LingMan** [comment]
@ncw Thanks for the ping.

The problem is that `walk.ListR` processes the entries of a folder in batches of 100. In my case the duplicate folders happen two be split across two different batches and thus don't get detected as dupes.
In contrast, `walk.Walk` calls the callback just once but with all 980 subfolders in the `entries` array.

**Comment by ncw** [comment]
> The problem is that `walk.ListR` processes the entries of a folder in batches of 100. In my case the duplicate folders happen to be split across two different batches and thus don't get detected as dupes.
> In contrast, `walk.Walk` calls the callback just once but with all 980 subfolders in the `entries` array.

Duh, you are right!  Thanks for pointing this out.

I've merged a fix for this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.47

Please can you give it a go?

**Comment by LingMan** [comment]
Yes, this works. Thank you.
